bentoml/repository/metadata_store.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2019 Atalaya Tech, Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import logging
+import datetime
+
+from sqlalchemy import (
+    Column,
+    Enum,
+    String,
+    Integer,
+    JSON,
+    Boolean,
+    DateTime,
+    UniqueConstraint,
+)
+from sqlalchemy.orm.exc import NoResultFound
+from google.protobuf.json_format import MessageToDict, ParseDict
+
+from bentoml.exceptions import BentoMLRepositoryException
+from bentoml.db import Base, create_session
+from bentoml.proto.repository_pb2 import (
+    UploadStatus,
+    BentoUri,
+    BentoServiceMetadata,
+    Bento as BentoPB,
+)
+
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_UPLOAD_STATUS = UploadStatus(status=UploadStatus.UNINITIALIZED)
+
+
+class Bento(Base):
+    __tablename__ = 'bentos'
+    __table_args__ = tuple(UniqueConstraint('name', 'version', name='_name_version_uc'))
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String, nullable=False)
+    version = Column(String, nullable=False)
+
+    # Storage URI for this Bento
+    uri = Column(String, nullable=False)
+    uri_type = Column(Enum(*BentoUri.StorageType.keys()), default=BentoUri.UNSET)
+
+    # JSON filed mapping directly to BentoServiceMetadata proto message
+    bento_service_metadata = Column(JSON, nullable=False, default={})
+
+    # Time of AddBento call, the time of Bento creation can be found in metadata field
+    created_at = Column(DateTime, default=datetime.datetime.utcnow)
+
+    # latest upload status, JSON message also includes last update timestamp
+    upload_status = Column(
+        JSON, nullable=False, default=MessageToDict(DEFAULT_UPLOAD_STATUS)
+    )
+
+    # mark as deleted
+    deleted = Column(Boolean, default=False)
+
+
+def _bento_orm_obj_to_pb(bento_obj):
+    bento_service_metadata_pb = ParseDict(
+        bento_obj.bento_service_metadata, BentoServiceMetadata()
+    )
+    bento_uri = BentoUri(
+        uri=bento_obj.uri, uri_type=BentoUri.StorageType.Value(bento_obj.uri_type)
+    )
+    return BentoPB(
+        name=bento_obj.name,
+        version=bento_obj.version,
+        uri=bento_uri,
+        bento_service_metadata=bento_service_metadata_pb,
+    )
+
+
+class BentoMetadataStore(object):
```

Comment:
Maybe because I was just reading protobuf stuff on Bento and BentoMetadata in repository.proto, the naming `BentoMetadataStore` is bit confusing.  Maybe to `BentoStore`?  I am not exactly sure.

Reply:
probably gonna merge it with the BentoRepository concept in follow up PR